Objects/descrobject.c

```diff
@@ -314,21 +312,20 @@ classmethoddescr_call(PyMethodDescrObject *descr, PyObject *args,
     self = PyTuple_GET_ITEM(args, 0);
     if (!PyType_Check(self)) {
         PyErr_Format(PyExc_TypeError,
-                     "descriptor '%V' requires a type "
-                     "but received a '%.100s'",
+                     "descriptor '%V' for type '%.100s' "
+                     "needs a type, not a '%.100s'",
                      descr_name((PyDescrObject *)descr), "?",
                      PyDescr_TYPE(descr)->tp_name,
                      self->ob_type->tp_name);
         return NULL;
     }
     if (!PyType_IsSubtype((PyTypeObject *)self, PyDescr_TYPE(descr))) {
         PyErr_Format(PyExc_TypeError,
-                     "descriptor '%V' "
-                     "requires a subtype of '%.100s' "
-                     "but received '%.100s",
+                     "descriptor '%V' for type '%.100s' "
+                     "doesn't apply to a '%.100s' type",
                      descr_name((PyDescrObject *)descr), "?",
                      PyDescr_TYPE(descr)->tp_name,
-                     self->ob_type->tp_name);
+                     ((PyTypeObject*)self)->tp_name);
```

Comment:
Maybe, this is a really old bug [commit](https://github.com/python/cpython/commit/7295c6a871f9cc42a4209a8eebe2e0974194a2a3).

```
>>> desc = set.add
>>> desc(int)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: descriptor 'add' requires a 'set' object but received a 'type'
```

This should be `...but received a 'int'`, not 'type'.

Reply:
```
>>> desc(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: descriptor 'add' requires a 'set' object but received a 'int'
>>> desc(set)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: descriptor 'add' requires a 'set' object but received a 'type'
```

Reply:
This does not look like a bug to me. If you will change it to produce `...but received a 'int'`, you will get `... requires a 'set' object but received a 'set'` for `desc(set)`, and I have no idea what you will get for `descr(1)`.

Reply:
I'm sorry, my example was wrong.  This is classmethoddesc_call.

```
>>> desc2 = dict.__dict__['fromkeys']
>>> desc2(int, [])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: descriptor 'fromkeys' requires a subtype of 'dict' but received 'type
```

`'type` must be `'int'`.

Reply:
filed https://bugs.python.org/issue36433